server/routerlicious/packages/services-core/src/dbFactoryFactory.ts

```diff
@@ -0,0 +1,59 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { Provider } from "nconf";
+import registerDebug from "debug";
+import { IDbFactory } from "./database";
+
+export interface IBackendDescriptor {
+    name: string // name of the backend
+    factory: (backendName?: string) => Promise<IDbFactory> // name of the constructor
+}
+
+export interface IBackendConfig {
+    path: string, // Path of the module to load
+    config: any, // configuration that will be passed into the constructor
+    name: string // name of the backend
+    factory: string // name of the constructor
+}
+const creator = (descriptor: IBackendConfig) => async () => {
+    debug(`Using ${descriptor.name} Backend`);
+    const extension = await import(`${descriptor.path}`);
+    const thingyFactory = extension[descriptor.factory];
+    return new thingyFactory(descriptor.config) as IDbFactory;
+};
+
+const debug = registerDebug("fluid:backend");
+
+export class DbFactoryFactory {
+    private readonly backends: Map<string, () => Promise<IDbFactory>>;
+
+    constructor(config: Provider, dbServices: IBackendDescriptor[], private readonly defaultBackend: string) {
+        const available_backends: Map<string, () => Promise<IDbFactory>> = new Map(
+            dbServices.map((desc) => [desc.name, desc.factory]),
+        );
+
+        if (config.get('loadExtensions')) {
```

Comment:
I think this change would overall make more sense if, instead of using a super generic concept like "extensions," we were to just make a single "db" config. Perhaps I am missing something, but I don't see a need for having the additional complexity of loading and configuring several database modules.

Reply:
Yes I agree that it adds additional complexity. 

The idea is to provide some flexibility for the users, to load custom DB drivers which are not part of the Fluid repo (e.g. DynamoDB driver). Additionally, to introduce some infrastructure in place to allow adding other extensions in the future, which are not necessary for db.